beacon-chain/rpc/service.go

```diff
@@ -238,7 +243,91 @@ func (s *Service) LatestCrystallizedState(req *empty.Empty, stream pb.BeaconServ
 // ValidatorAssignment streams validator assignments every slot to clients that request
 // to watch a subset of public keys in the CrystallizedState's active validator set.
 func (s *Service) ValidatorAssignment(req *pb.ValidatorAssignmentRequest, stream pb.ValidatorService_ValidatorAssignmentServer) error {
-	return nil
+	// If the genesis time was at 12:00:00PM a if the SlotDuration is set to
+	// 8 seconds. We can accomplish this by using utils.WaitUntilTimestamp
+	// and passing in the desired slot duration as an argument.
+	//
+	// Instead of utilizing params.SlotDuration, we utilize a property of
+	// RPC service struct called slotAlignmentDuration. This value
+	// can be set to 0 seconds as a parameter in tests.
+	// Otherwise, tests would sleep.
+	utils.WaitUntilTimestamp(time.Duration(s.slotAlignmentDuration) * time.Second)
+
+	// Right after the timestamps are aligned, we start a new ticker
+	// that fires exactly every params.SlotDuration*time.Second. This ensures
+	// a consistent "heartbeat" to keep track of every beacon slot.
+	// NOTE: Change to a parameter as well.
+	tickerChan := time.NewTicker(time.Duration(params.SlotDuration) * time.Second).C
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Debug("ValidatorAssignment stream closed, exiting goroutine")
+			return nil
+		case <-tickerChan:
+			log.WithField("slotNumber", utils.CurrentBeaconSlot()).Debug("Sending validator assignment to RPC clients")
+			// NOTE: this set will not change, so we can just cache the previous one
+			// we computed and stream instead of recomputing each slot.
+
+			// We get the current crystallized state.
+			cState := s.chainService.CurrentCrystallizedState()
+
+			// Next, for each public key in the request, we build
+			// up an array of assignments.
+			assignments := []*pb.ValidatorAssignmentResponse_Assignment{}
+			for _, val := range req.GetPublicKeys() {
+
+				// For the corresponding public key and current crystallized state,
+				// we determine the assigned slot for the validator and whether it
+				// should act as a proposer or attester.
+				assignedSlot, responsibility, err := casper.ValidatorSlotAndResponsibility(
+					val.GetPublicKey(),
+					cState.CurrentDynasty(),
+					cState.Validators(),
+					cState.ShardAndCommitteesForSlots(),
+				)
+				if err != nil {
+					return err
+				}
+
+				var role pb.ValidatorRole
+				if responsibility == "proposer" {
+					role = pb.ValidatorRole_PROPOSER
+				} else {
+					role = pb.ValidatorRole_ATTESTER
+				}
+
+				// We determine the assigned shard ID for the validator
+				// based on a public key and current crystallized state.
+				shardID, err := casper.ValidatorShardID(
```

Comment:
Another thought, should we just combine `ValidatorShardID` with `ValidatorSlotAndResponsibility`? That way ValidatorSlotAndResponsibility returns `slot, shard id and responsibility`?

Reply:
I can open a PR for you on that if you want